### PR TITLE
[WIP] Adds combat system

### PR DESCRIPTION
**Pull request by Qustinnus**
far from done but PRing for visibility and discussion

### CODE REVIEW

**Comment by Keekenox** [comment]
Shit PR.

**Comment by Qustinnus** [comment]
hi keek

**Comment by Keekenox** [comment]
I love you, Floyd.

**Comment by Kmc2000** [comment]
nice sprite credit chumbis i'll eat you alive

**Comment by Qustinnus** [comment]
i havnt even added a cl yet chesse20 looking motherfucker

**Comment by Qustinnus** [comment]
I'm waiting for TMTime's PR then I'll start implementing more combat stuff 👀

**Comment by Dennok** [comment]
I don't even, but i think TMTime wait you.
@TMTIME what you think about this?

**Comment by Qustinnus** [comment]
no because his systems dont rely on mine, but my systems do rely on his.

**Comment by TMTIME** [comment]
Yeah I'm kinda waiting for mine to be merged. If you really want I can review this now @Qustinnus

**Comment by Qustinnus** [comment]
@tmtime yes feel free, the weapons themselves are WIP atm because im refactoring them a little bit but a lot of other stuff should be fine atm.

**Comment by Dennok** [comment]
Just lost here comment to have it in participating.

**Comment by TMTIME** [comment]
13 days late lets go

File: ftl13.dme [hunk 1]
```diff
@@ -15,4 +15,19 @@
...
\ No newline at end of file
+#include "ftl\code\controllers\subsystem\ships.dm"
+#include "ftl\code\datums\shipattacks\player_attacks\player_attack.dm"
+#include "ftl\code\datums\ships\shiproom.dm"
+#include "ftl\code\datums\ships\shiptemplates.dm"
+#include "ftl\code\datums\ships\starship.dm"
+#include "ftl\code\datums\ships\shiprooms\shiprooms.dm"
+#include "ftl\code\game\area\areas.dm"
+#include "ftl\code\game\area\shiprooms.dm"
+#include "ftl\code\game\machinery\shipweapon.dm"
```
[the remark this answers is not in the document]
Comment (Qustinnus) [reply ftl13.dme]: huh?

[the remark this answers is not in the document]
Comment (TMTIME) [reply ftl13.dme]: You moved the original file to ftl/code/game/machinery/weapons/shipweapon.dm

File: ftl/code/datums/ships/starship.dm [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	SSships.currentships -= unique_id
+	for(var/i in template.get_affected_turfs(ship_spawn_slot.loc, TRUE)) //this is so shit TODO: kill this unless this is our best way of cleaning up.
+		var/turf/T = i
+		for(var/x in T.contents)
+			qdel(x)
+		qdel(T)
+		CHECK_TICK
+
+/datum/starship/proc/adjust_hull(value) //use this to change hull level or i kill you
+	hull_integrity = max(hull_integrity + value, initial(hull_integrity))
```
Comment (TMTIME) [inline ftl/code/datums/ships/starship.dm]: If you use this to damage the ship, it will always pick the initial hull hp

**Comment by TMTIME** [comment]
also the physical shots come out going north

File: ftl/code/datums/ships/starship.dm (hunk 2 above)
Comment (Qustinnus) [reply ftl/code/datums/ships/starship.dm]: oh shit my bad this should be clamping thanks

**Comment by Qustinnus** [comment]
@TMTIME reviw pls. also I cant figure out why the shots come out facing north reeee

File: ftl/code/game/objects/items/weapon_chips/projectile_chip.dm [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+/obj/item/weapon_chip/projectile/proc/SpawnProjectile() //Projectile that flies out of the gun and dissapears, exists for visual aesthetic
+	var/obj/item/projectile/ship_projectile/A = new(weapon.loc)
+	A.icon_state =  projectile_icon
+	A.setDir(EAST)
+	A.pixel_x = 32
+	A.pixel_y = 12
+	A.yo = 0
+	A.xo = 20
+	A.starting = weapon
+	A.fire()
```
Comment (Dennok) [inline ftl/code/game/objects/items/weapon_chips/projectile_chip.dm]:
```suggestion
	A.fire(dir2angle(weapon.dir))
```
This fix fired projectile direction

**Comment by Dennok** [comment]
Someone who first make ship weapon take emitter, and make it to fire on button click. 
I think now we need to make it again.

File: ftl/code/datums/ships/starship.dm [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	var/combat_map = "combat/generic_ship.dmm"
+
+/datum/starship/New(turf/T, var/ship_spawn_slot)
+	. = ..()
+	unique_id = "[name] [++ship_count]"
+	SSships.currentships[unique_id] = src
+
+	var/map = "[prefix][combat_map]"
+	template = new(map)
+	template.load(T, TRUE)
```
Comment (Dennok) [inline ftl/code/datums/ships/starship.dm]:
```suggestion
	var/list/bounds = template.load(T, TRUE)

	var/list/turfs = block(	locate(bounds[MAP_MINX], bounds[MAP_MINY], bounds[MAP_MINZ]),
							locate(bounds[MAP_MAXX], bounds[MAP_MAXY], bounds[MAP_MAXZ]))
```

File: ftl/code/datums/ships/starship.dm [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+	src.ship_spawn_slot = ship_spawn_slot
+	SSships.ShipSpawnLocations[ship_spawn_slot] = FALSE //This slot is taken now, cya chump.
+
+	var/shipareas
+	for(var/turf/open/indestructible/ftlfloor/floor in template.get_affected_turfs(src.ship_spawn_slot.loc))
```
Comment (Dennok) [inline ftl/code/datums/ships/starship.dm]:
```suggestion
	for(var/turf/open/indestructible/ftlfloor/floor in turfs)
```

File: ftl/code/datums/ships/starship.dm [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+		floor.unique_id = unique_id
+
+		var/area/ftl/shiproom/A = get_area(floor)
+		if(!shiprooms.Find(A))
```
Comment (Dennok) [inline ftl/code/datums/ships/starship.dm]:
```suggestion
		if(shiprooms.Find(A))
```

**Comment by crossedfall** [comment]
Closing this PR and moving discussion to the new PR: #657

**Pull request closed without merging**